### PR TITLE
Add Annotations for config validation checking

### PR DESCRIPTION
**Pull request by srkukarni**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

This pr adds annotations that config classes can use to do member validation. This will be useful for things like source/sink/function submission to check if supplied config is in order before starting the actual source/sink/function

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                if (!value.startsWith("ABCDE")) {
+                    throw new IllegalArgumentException(String.format("Field %s does not start with ABCDE", name));
+                }
+            } else {
+                throw new IllegalArgumentException(String.format("Field %s is not a string", name));
+            }
+        }
+    }
+
+    class TestConfig {
```
Comment (cckellogg) [inline pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: I like the idea of having type checking. The naming is a bit confusing to me something like makes more sense to me. 
```
class TestConfig {
       @NotNull
        public String stringValue;
        
       @PositiveNumber or @Number(positive=true)
        public Integer positiveNumber;
        
       @List(itemType = Integer.class) or @ListType(itemType = Integer.class) or @IntegerList
        public List integerList;
        
       @Map(keyType = String.class, valueType = Integer.class) or @MapType(keyType = String.class, valueType = Integer.class)
        public Map stringIntegerMap;
        
       @List(itemType = String.class) or @StringList
        public List stringList;
        
       @TopicName
        public String topic;
        
       @CustomType(validatorClass = TestValidator.class)
        public String customString;
    }
```
I think annotations should describe what the variable type is as opposed to describing how to test it. The testing is implied. 

Another thing to think about is how to handle fields that should be required to have a value.

Thoughts?

Thoughts? Al

Comment (jerrypeng) [reply pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: I don't fee strongly about it either way.   However, if we feel we can simplify the naming of annotations and still be intuitive to the end user, we should.  The current annotations are borrowed from Storm that I created a while ago:

https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/validation/ConfigValidationAnnotations.java

Others have name annotations in a more simplified way:

https://www.baeldung.com/javax-validation

Comment (srkukarni) [reply pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: I have changed the annotations. wrt fields that should have a value, isn't that covered by NotNull?

Comment (cckellogg) [reply pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: Can NotNull be applied in combination with another annotation like this? Or does an annotation like `@TopicName` or `@List` imply not null?
```
@NotNull
@TopicName
 public String topic;
```

Comment (srkukarni) [reply pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: Multiple annotations are possible on a field.
Thus
@NotNull
@TopicName implies that it should be non null and is a topic. However I believe that TopicName already ensures that its not null

Comment (cckellogg) [reply pulsar-common/src/test/java/org/apache/pulsar/common/validator/ConfigValidationTest.java:58]: Sounds good and LGTM +1

**Pull request merged**